src/icf/interfaceCallForwarder.ts

```diff
@@ -0,0 +1,463 @@
+ * @param filePath : path to the cairo contract file
+ * @param options : options for the generation (solc version, output directory, etc.)
+ * @returns : none
+ */
+export function generateSolInterface(filePath: string, options: SolcInterfaceGenOptions) {
+  const cairoPathRoot = filePath.slice(0, -'.cairo'.length);
+  const { success } = compileCairo(filePath, path.resolve(__dirname, '..'), {
+    debugInfo: false,
+  });
+
+  // files generated by the compilation of the given cairo contract
+  const resultPath = `${cairoPathRoot}_compiled.json`;
+  const abiPath = `${cairoPathRoot}_abi.json`;
+
+  if (!success) {
+    logError(`Compilation of contract ${filePath} failed`);
+    return;
+  } else {
+    fs.unlinkSync(resultPath);
+  }
+
+  let solPath = `${cairoPathRoot}.sol`; // default path for the generated solidity file
+  let cairoContractPath = `${cairoPathRoot}_forwarder.cairo`; // default path for the generated cairo contract file
+
+  if (options.output) {
+    const filePathRoot = options.output.slice(0, -'.sol'.length);
+    solPath = options.output;
+    cairoContractPath = filePathRoot + '.cairo';
+  }
+
+  const abi: AbiType = JSON.parse(fs.readFileSync(abiPath, 'utf8'));
+
+  // generate the cairo contract that will be used to interact with the given cairo contract
+  let cairoContract: string = genCairoContract(abi, options.contractAddress, options.classHash);
+
+  const writer = new ASTWriter(
+    new Map<ASTNodeConstructor<ASTNode>, ASTNodeWriter>([...DefaultASTWriterMapping]),
+    new PrettyFormatter(4, 0),
+    options.solcVersion ?? defaultSolcVersion,
+  );
+
+  const sourceUint = new SourceUnit(0, '', solPath, 0, solPath, new Map(), []);
+  sourceUint.context = new ASTContext(sourceUint);
+  const ast = new AST([sourceUint], options.solcVersion ?? defaultSolcVersion, {
+    contracts: {},
+    sources: {},
+  });
+
+  addPragmaDirective(options.solcVersion ?? defaultSolcVersion, sourceUint, ast);
+
+  // add uint256 version of the structs from the cairo contract after transformation
+  // of all felt types to uint256 to the source unit and also the structs that represents
+  // the heterogeneous tuples used in the given cairo contract
+  const structDefs: Map<string, StructDefinition> = addAllStructs(sourceUint, ast, abi);
+
+  // solidity interface contract
+  const { contract, funcSignatures } = addForwarderContract(
+    abi,
+    sourceUint,
+    structDefs,
+    ast,
+    cairoPathRoot.split('/').pop(),
+  );
+
+  fs.unlinkSync(abiPath);
+
+  // replace function names in the generated cairo contract such that it'll match the function
+  // names in the generated solidity contract after transpilation of the latter
+  funcSignatures.forEach((value, key) => {
+    cairoContract = cairoContract.replace(`_ITR_${key}`, `${key}_${value}`);
+  });
+
+  fs.writeFileSync(cairoContractPath, cairoContract);
+
+  const compileForwarder = compileCairo(cairoContractPath, path.resolve(__dirname, '..'), {
```

Comment:
what are the cases when our generated cairo contract will fail? Can we remove this second compileCairo call?

Reply:
It is needed to do for generation of  class_hash